### ISSUE TITLE
Fixed reference counting issues

### ISSUE THREAD

**Issue by RobAustin** [OpenHFT/Chronicle-Core#70]
10:27:38.985 [main] INFO net.openhft.chronicle.queue.DirectoryUtils - Tmp dir: /Users/robaustin/git-projects/Chronicle-Queue/target/rollCycleStress-jia20sb6
Queue dir: /Users/robaustin/git-projects/Chronicle-Queue/target/rollCycleStress-jia20sb6 at 2018-06-11T09:27:38.985Z
Running test with 2 writers and 2 readers, sleep 50000ns
Writing 40000 messages with 50000ns interval
Should take ~1sec
```
java.lang.Throwable: 6e1ca577-main/reader-1 creation ref-count=1
	at net.openhft.chronicle.core.ReferenceCounter.newRefCountHistory(ReferenceCounter.java:45)
	at net.openhft.chronicle.core.ReferenceCounter.<init>(ReferenceCounter.java:35)
	at net.openhft.chronicle.core.ReferenceCounter.onReleased(ReferenceCounter.java:40)
	at net.openhft.chronicle.bytes.AbstractBytes.<init>(AbstractBytes.java:38)
	at net.openhft.chronicle.bytes.MappedBytes.<init>(MappedBytes.java:59)
	at net.openhft.chronicle.bytes.MappedBytes.<init>(MappedBytes.java:55)
	at net.openhft.chronicle.bytes.MappedBytes.mappedBytes(MappedBytes.java:113)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueStore.bytes(SingleChronicleQueueStore.java:358)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.resetWires(SingleChronicleQueueExcerpts.java:1745)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.cycle(SingleChronicleQueueExcerpts.java:1950)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.nextIndexWithNextAvailableCycle0(SingleChronicleQueueExcerpts.java:1547)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.nextIndexWithNextAvailableCycle(SingleChronicleQueueExcerpts.java:1508)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.endOfCycle(SingleChronicleQueueExcerpts.java:1294)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.next0(SingleChronicleQueueExcerpts.java:1269)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:1197)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:1129)
	at net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest$Reader.call(RollCycleMultiThreadStressTest.java:270)
	at net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest$Reader.call(RollCycleMultiThreadStressTest.java:237)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.Throwable: 6e1ca577-main/reader-1 Release ref-count=0
	at net.openhft.chronicle.core.ReferenceCounter.recordRelease(ReferenceCounter.java:88)
	at net.openhft.chronicle.core.ReferenceCounter.release(ReferenceCounter.java:79)
	at net.openhft.chronicle.bytes.AbstractBytes.release(AbstractBytes.java:469)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.close(SingleChronicleQueueExcerpts.java:1135)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue.lambda$close$4(SingleChronicleQueue.java:537)
	at java.util.WeakHashMap.forEach(WeakHashMap.java:1025)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue.close(SingleChronicleQueue.java:537)
	at net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest$Reader.call(RollCycleMultiThreadStressTest.java:302)
	at net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest$Reader.call(RollCycleMultiThreadStressTest.java:237)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.Throwable: 6e1ca577-main/queue-thread-local-cleaner-daemon Release ref-count=0
	at net.openhft.chronicle.core.ReferenceCounter.recordRelease(ReferenceCounter.java:88)
	at net.openhft.chronicle.core.ReferenceCounter.release(ReferenceCounter.java:74)
	at net.openhft.chronicle.bytes.AbstractBytes.release(AbstractBytes.java:469)
	at net.openhft.chronicle.queue.impl.single.StoreComponentReferenceHandler.processWireQueue(StoreComponentReferenceHandler.java:74)
	at net.openhft.chronicle.queue.impl.single.StoreComponentReferenceHandler.lambda$static$0(StoreComponentReferenceHandler.java:43)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.Throwable: 21bb9e0f-main/reader creation ref-count=1
	at net.openhft.chronicle.core.ReferenceCounter.newRefCountHistory(ReferenceCounter.java:45)
	at net.openhft.chronicle.core.ReferenceCounter.<init>(ReferenceCounter.java:35)
	at net.openhft.chronicle.core.ReferenceCounter.onReleased(ReferenceCounter.java:40)
	at net.openhft.chronicle.bytes.AbstractBytes.<init>(AbstractBytes.java:38)
	at net.openhft.chronicle.bytes.MappedBytes.<init>(MappedBytes.java:59)
	at net.openhft.chronicle.bytes.MappedBytes.<init>(MappedBytes.java:55)
	at net.openhft.chronicle.bytes.MappedBytes.mappedBytes(MappedBytes.java:113)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueStore.bytes(SingleChronicleQueueStore.java:358)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.resetWires(SingleChronicleQueueExcerpts.java:1745)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.cycle(SingleChronicleQueueExcerpts.java:1950)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.nextIndexWithNextAvailableCycle0(SingleChronicleQueueExcerpts.java:1547)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.nextIndexWithNextAvailableCycle(SingleChronicleQueueExcerpts.java:1508)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.endOfCycle(SingleChronicleQueueExcerpts.java:1294)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.next0(SingleChronicleQueueExcerpts.java:1269)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:1197)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:1129)
	at net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest$Reader.call(RollCycleMultiThreadStressTest.java:270)
	at net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest$Reader.call(RollCycleMultiThreadStressTest.java:237)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.Throwable: 21bb9e0f-main/reader Release ref-count=0
	at net.openhft.chronicle.core.ReferenceCounter.recordRelease(ReferenceCounter.java:88)
	at net.openhft.chronicle.core.ReferenceCounter.release(ReferenceCounter.java:79)
	at net.openhft.chronicle.bytes.AbstractBytes.release(AbstractBytes.java:469)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.close(SingleChronicleQueueExcerpts.java:1135)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue.lambda$close$4(SingleChronicleQueue.java:537)
	at java.util.WeakHashMap.forEach(WeakHashMap.java:1025)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue.close(SingleChronicleQueue.java:537)
	at net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest$Reader.call(RollCycleMultiThreadStressTest.java:302)
	at net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest$Reader.call(RollCycleMultiThreadStressTest.java:237)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.Throwable: 21bb9e0f-main/queue-thread-local-cleaner-daemon Release ref-count=0
	at net.openhft.chronicle.core.ReferenceCounter.recordRelease(ReferenceCounter.java:88)
	at net.openhft.chronicle.core.ReferenceCounter.release(ReferenceCounter.java:74)
	at net.openhft.chronicle.bytes.AbstractBytes.release(AbstractBytes.java:469)
	at net.openhft.chronicle.queue.impl.single.StoreComponentReferenceHandler.processWireQueue(StoreComponentReferenceHandler.java:74)
	at net.openhft.chronicle.queue.impl.single.StoreComponentReferenceHandler.lambda$static$0(StoreComponentReferenceHandler.java:43)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
10:27:40.082 [main] INFO net.openhft.chronicle.queue.DirectoryUtils - Tmp dir: /Users/robaustin/git-projects/Chronicle-Queue/target/rollCycleStress-jia20sb7
All messages written in 1secs at rate of 36,563/sec 18,282/sec per writer (actual writeLatency 54,700ns)
```
Test complete

**Comment by RobAustin** [comment]
@NickWard fyi

**Comment by hft-team-city** [comment]
Released in [Chronicle-Core-2.20.102](https://github.com/OpenHFT/Chronicle-Core/releases/chronicle-core-2.20.102), [BOM-2.20.134](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.134)
